File: Minimize_the_Maximum_Difference_of_Pairs.java
class Solution {
    public int minimizeMax(int[] nums, int p) {
        Arrays.sort(nums);

        int left = 0;
        int right = nums[nums.length - 1] - nums[0];

        while (left < right) {
            int midDiff = (left + right) / 2;
            if (isFeasible(midDiff, nums, p)) {
                right = midDiff;
            } else {
                left = midDiff + 1;
            }
        }

        return left;        
    }

    private boolean isFeasible(int diff, int[] nums, int p) {
        int count = 0;
        int i = 0;
        while (i < nums.length - 1 && count < p) {
            if (nums[i + 1] - nums[i] <= diff) {
                count++;
                i += 2;
            } else {
                i++;
            }
        }
        return count >= p;
    }    
}

File: Minimize_the_Maximum_Difference_of_Pairs.py
class Solution:
    def minimizeMax(self, nums: List[int], p: int) -> int:
        nums.sort()

        def isFeasible(diff):
            count = 0
            i = 0
            while i < len(nums) - 1 and count < p:
                if nums[i+1] - nums[i] <= diff:
                    count += 1
                    i += 2
                else:
                    i += 1
            
            return count >= p

        left, right = 0, nums[-1] - nums[0]

        while left < right:
            mid_diff = (left + right) // 2
            if isFeasible(mid_diff):
                right = mid_diff
            else:
                left = mid_diff + 1
        
        return left
